[PATCH] SBV library: remove libc dependency

This patch eliminates the only dependency from libpatches to libc, so
libpatches.erl only depends on libkernel.erl.

Signed-off-by: jimmikaelkael <jimmikaelkael@wanadoo.fr>
Signed-off-by: misfire <misfire@xploderfreax.de>

diff -r 40ac7ccfd3f0 ee/sbv/src/erl-support.c
--- a/ee/sbv/src/erl-support.c	Tue Sep 01 15:50:30 2009 +0200
+++ b/ee/sbv/src/erl-support.c	Mon Sep 07 13:49:46 2009 +0200
@@ -17,6 +17,5 @@
 
 char * erl_dependancies[] = {
     "libkernel",
-    "libc",
     0
 };
diff -r 40ac7ccfd3f0 ee/sbv/src/slib.c
--- a/ee/sbv/src/slib.c	Tue Sep 01 15:50:30 2009 +0200
+++ b/ee/sbv/src/slib.c	Mon Sep 07 13:49:46 2009 +0200
@@ -22,6 +22,20 @@
 #define SEARCH_SIZE	(16 * 1024)
 static u8 smem_buf[SEARCH_SIZE];
 
+/* Do not link to memcmp() from libc, so we only depend on libkernel. */
+int __memcmp(const void *s1, const void *s2, unsigned int length)
+{
+	const char *a = s1;
+	const char *b = s2;
+
+	while (length--) {
+		if (*a++ != *b++)
+			return 1;
+	}
+
+	return 0;
+}
+
 /**
  * slib_exp_lib_list - Find the head and tail of the export library list.
  *
@@ -57,7 +71,7 @@
 		/* SYSMEM's export library sits at 0x830, so it should appear in
 		   LOADCORE's prev pointer.  */
 		if (*(u32 *)(smem_loc + i) == 0x830) {
-			if (!memcmp(smem_loc + i + 12, "loadcore", 8))
+			if (!__memcmp(smem_loc + i + 12, "loadcore", 8))
 				break;
 		}
 	}
@@ -105,7 +119,7 @@
 	while (cur_lib) {
 		smem_read(cur_lib, exp_lib, sizeof buf);
 
-		if (!memcmp(exp_lib->name, name, len)) {
+		if (!__memcmp(exp_lib->name, name, len)) {
 			while (exp_lib->exports[count] != 0)
 				count++;
 
diff -r 40ac7ccfd3f0 ee/sbv/src/smod.c
--- a/ee/sbv/src/smod.c	Tue Sep 01 15:50:30 2009 +0200
+++ b/ee/sbv/src/smod.c	Mon Sep 07 13:49:46 2009 +0200
@@ -17,6 +17,9 @@
 #include "smem.h"
 #include "smod.h"
 
+/* from slib.c */
+extern int __memcmp(const void *s1, const void *s2, unsigned int length);
+
 /**
  * smod_get_next_mod - Return the next module referenced in the global module list.
  */
@@ -52,7 +55,7 @@
 	do {
 		smem_read(info->name, search_name, sizeof search_name);
 
-		if (!memcmp(search_name, name, len))
+		if (!__memcmp(search_name, name, len))
 			return info->id;
 	} while (smod_get_next_mod(info, info) != 0);
 
